core/src/test/scala/spinal/core/sim/SpinalFunSuite.scala

```diff
@@ -10,6 +10,99 @@ class Identity extends Component {
   io.output := io.input
 }
 
+class BadIdentity extends Identity {
+  io.output := io.output
+}
+
+
+// Drawback: I have to provide a base type as I do not take arguments anymore
+class IdentityFlatSpec extends SpinalFlatSpec[Identity] {
+
+  // Global parameters
+  override val config = SimConfig.withWave
+  // override val caching = false
+
+  // Parameters can be erased by local parameter when building benches
+  val asIs = bench(new Identity, caching = false)
+  // I can transform, and the "software abstraction" type is inferred
+  val transformed = benchTransform(new Identity, dut => SoftIdentity(dut))
+  // These ones will not compile
+  val failing = bench(new BadIdentity)
+  val failingNoCache = bench(new BadIdentity, caching = false)
+
+  // Before my first test, it tests if it compiles, and cancels test if it doesn't
+  // Here I provide an empty test and I get:
+  //   failing should compile *** FAILED ***
+  //   failing should do nothing !!! CANCELED !!!
+  failing should "do nothing" in { dut => }
+
+  //    failingNoCache should do nothing *** FAILED ***
+  failingNoCache should "do nothing" in { dut => }
+
+  // Notice that benches are `Nameable` (but I did not implement recursive names)
+
+  // Here I have no caching so no "compiles" test, I get:
+  //   asIs should pass random test
+  asIs should "pass random tests" in { dut =>
+    for (i <- 0 to 99) {
+      val v = dut.io.input.randomize()
+      sleep(1)
+      assert(dut.io.output.toInt == v)
+    }
+  }
+
+  //   asIs should pass all tests
+  asIs should "pass all tests" in { dut =>
+    for (v <- 0 to 255) {
+      dut.io.input #= v
+      sleep(1)
+      assert(dut.io.output.toInt == v)
+    }
+  }
+
+  //   transformed should compile
+  //   transformed should pass random tests
+  transformed should "pass random tests" in { dut =>
+    for (i <- 0 to 99) {
+      val v = dut.dut.io.input.randomizedInt()
+      assert(dut(v) == v)
+    }
+  }
+
+  //   transformed should pass all tests
+  transformed should "pass all tests" in { dut =>
+    for (v <- 0 to 255)
+      assert(dut(v) == v)
+  }
+}
+
+
+class IdentityFlatSpecForLoop extends SpinalFlatSpec[Identity] {
+  override val config = SimConfig.withWave
+
+  val asIs = bench(new Identity, caching = false)
+  val failing = bench(new BadIdentity)
+  val failingNoCache = bench(new BadIdentity, caching = false)
+
+  for (b <- Seq(asIs, failing, failingNoCache)) {
```

Comment:
```
IdentityFlatSpecForLoop:
asIs
- should pass random tests
asIs
- should pass all tests
failing
- should compile *** FAILED ***
failing
- should pass random tests !!! CANCELED !!!
failing
- should pass all tests !!! CANCELED !!!
failingNoCache
- should pass random tests *** FAILED ***
failingNoCache
- should pass all tests *** FAILED ***
Execution took 17s
7 tests, 2 passed, 3 failed, 2 canceled

Failed:
- spinal.core.sim.IdentityFlatSpecForLoop:
  * failing should pass random tests
  * failingNoCache should pass random tests - spinal.core.SpinalExit: 
  * failing should compile - spinal.core.SpinalExit: 
  * failing should pass all tests
  * failingNoCache should pass all tests - spinal.core.SpinalExit: 
```

<details><summary>Unfiltered result</summary>

Note: it might be good to enable HTML result for the users XD

```
IdentityFlatSpecForLoop:
asIs
[Runtime] SpinalHDL dev    git head : f2906310aff8111311df2061920f4fb1e8be7ee4
[Runtime] JVM max memory : 3968,0MiB
[Runtime] Current date : 2022.11.17 16:24:14
[Progress] at 0,000 : Elaborate components
[Progress] at 0,323 : Checks and transforms
[Progress] at 0,504 : Generate Verilog
[Done] at 0,584
[Progress] Simulation workspace in /home/come/Development/SpinalHDL/core/./simWorkspace/Identity
[Progress] Verilator compilation started
[Progress] Verilator compilation done in 9897,220 ms
[Progress] Start Identity asIs should pass random tests simulation with seed 0
[Done] Simulation done in 85,083 ms
- should pass random tests
asIs
[Runtime] SpinalHDL dev    git head : f2906310aff8111311df2061920f4fb1e8be7ee4
[Runtime] JVM max memory : 3968,0MiB
[Runtime] Current date : 2022.11.17 16:24:25
[Progress] at 10,658 : Elaborate components
[Progress] at 10,682 : Checks and transforms
[Progress] at 10,686 : Generate Verilog
[Done] at 10,735
[Info] Workspace 'Identity' was reallocated as 'Identity_1' to avoid collision
[Progress] Simulation workspace in /home/come/Development/SpinalHDL/core/./simWorkspace/Identity_1
[Progress] Verilator compilation started
[info] Found cached verilator binaries
[Progress] Verilator compilation done in 6301,111 ms
[Progress] Start Identity asIs should pass all tests simulation with seed 0
[Done] Simulation done in 35,041 ms
- should pass all tests
failing
[Runtime] SpinalHDL dev    git head : f2906310aff8111311df2061920f4fb1e8be7ee4
[Runtime] JVM max memory : 3968,0MiB
[Runtime] Current date : 2022.11.17 16:24:31
[Progress] at 17,101 : Elaborate components
[Progress] at 17,116 : Checks and transforms

**********************************************************************************************
[Warning] Elaboration failed (2 errors).
          Spinal will restart with scala trace to help you to find the problem.
**********************************************************************************************

[Progress] at 17,128 : Elaborate components
[Progress] at 17,157 : Checks and transforms
- should compile *** FAILED ***
  spinal.core.SpinalExit: Error detected in phase PhaseCheck_noLatchNoOverride
********************************************************************************
********************************************************************************

ASSIGNMENT OVERLAP completely the previous one of (toplevel/io_output : out Bits[8 bits])
    spinal.sim.JvmThread.run(SimManager.scala:51)


********************************************************************************
********************************************************************************

Design's errors are listed above.
SpinalHDL compiler exit stack :
  at spinal.core.SpinalExit$.apply(Misc.scala:424)
  at spinal.core.SpinalError$.apply(Misc.scala:479)
  at spinal.core.internals.PhaseContext.checkPendingErrors(Phase.scala:175)
  at spinal.core.internals.PhaseContext.doPhase(Phase.scala:191)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2$$anonfun$apply$137.apply(Phase.scala:2707)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2$$anonfun$apply$137.apply(Phase.scala:2705)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2.apply(Phase.scala:2705)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2.apply(Phase.scala:2641)
  ...
failing
- should pass random tests !!! CANCELED !!!
  org.scalatest.exceptions.TestCanceledException was thrown. (SpinalFunSuite.scala:62)
failing
- should pass all tests !!! CANCELED !!!
  org.scalatest.exceptions.TestCanceledException was thrown. (SpinalFunSuite.scala:62)
failingNoCache
[Runtime] SpinalHDL dev    git head : f2906310aff8111311df2061920f4fb1e8be7ee4
[Runtime] JVM max memory : 3968,0MiB
[Runtime] Current date : 2022.11.17 16:24:32
[Progress] at 17,255 : Elaborate components
[Progress] at 17,285 : Checks and transforms

**********************************************************************************************
[Warning] Elaboration failed (3 errors).
          Spinal will restart with scala trace to help you to find the problem.
**********************************************************************************************

[Progress] at 17,294 : Elaborate components
[Progress] at 17,315 : Checks and transforms
- should pass random tests *** FAILED ***
  spinal.core.SpinalExit: Error detected in phase PhaseCheck_noLatchNoOverride
********************************************************************************
********************************************************************************

ASSIGNMENT OVERLAP completely the previous one of (toplevel/io_output : out Bits[8 bits])
    spinal.sim.JvmThread.run(SimManager.scala:51)


********************************************************************************
********************************************************************************

Design's errors are listed above.
SpinalHDL compiler exit stack :
[Runtime] SpinalHDL dev    git head : f2906310aff8111311df2061920f4fb1e8be7ee4
[Runtime] JVM max memory : 3968,0MiB
[Runtime] Current date : 2022.11.17 16:24:32
[Progress] at 17,337 : Elaborate components
[Progress] at 17,356 : Checks and transforms

**********************************************************************************************
[Warning] Elaboration failed (3 errors).
          Spinal will restart with scala trace to help you to find the problem.
**********************************************************************************************

[Progress] at 17,361 : Elaborate components
[Progress] at 17,369 : Checks and transforms
  at spinal.core.SpinalExit$.apply(Misc.scala:424)
  at spinal.core.SpinalError$.apply(Misc.scala:479)
  at spinal.core.internals.PhaseContext.checkPendingErrors(Phase.scala:175)
  at spinal.core.internals.PhaseContext.doPhase(Phase.scala:191)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2$$anonfun$apply$137.apply(Phase.scala:2707)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2$$anonfun$apply$137.apply(Phase.scala:2705)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2.apply(Phase.scala:2705)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2.apply(Phase.scala:2641)
  ...
failingNoCache
- should pass all tests *** FAILED ***
  spinal.core.SpinalExit: Error detected in phase PhaseCheck_noLatchNoOverride
********************************************************************************
********************************************************************************

ASSIGNMENT OVERLAP completely the previous one of (toplevel/io_output : out Bits[8 bits])
    spinal.sim.JvmThread.run(SimManager.scala:51)


********************************************************************************
********************************************************************************

Design's errors are listed above.
SpinalHDL compiler exit stack :
  at spinal.core.SpinalExit$.apply(Misc.scala:424)
  at spinal.core.SpinalError$.apply(Misc.scala:479)
  at spinal.core.internals.PhaseContext.checkPendingErrors(Phase.scala:175)
  at spinal.core.internals.PhaseContext.doPhase(Phase.scala:191)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2$$anonfun$apply$137.apply(Phase.scala:2707)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2$$anonfun$apply$137.apply(Phase.scala:2705)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2.apply(Phase.scala:2705)
  at spinal.core.internals.SpinalVerilogBoot$$anonfun$singleShot$2.apply(Phase.scala:2641)
  ...
Execution took 17s
7 tests, 2 passed, 3 failed, 2 canceled

===============================================
Total duration: 17s
1 failed

Failed:
- spinal.core.sim.IdentityFlatSpecForLoop:
  * failing should pass random tests
  * failingNoCache should pass random tests - spinal.core.SpinalExit: 


 Error detected in phase PhaseCheck_noLatchNoOverride
********************************************************************************
********************************************************************************

ASSIGNMENT OVERLAP completely the previous one of (toplevel/io_output : out Bits[8 bits])
    spinal.sim.JvmThread.run(SimManager.scala:51)


********************************************************************************
********************************************************************************

Design's errors are listed above.
SpinalHDL compiler exit stack :
  * failing should compile - spinal.core.SpinalExit: 


 Error detected in phase PhaseCheck_noLatchNoOverride
********************************************************************************
********************************************************************************

ASSIGNMENT OVERLAP completely the previous one of (toplevel/io_output : out Bits[8 bits])
    spinal.sim.JvmThread.run(SimManager.scala:51)


********************************************************************************
********************************************************************************

Design's errors are listed above.
SpinalHDL compiler exit stack :
  * failing should pass all tests
  * failingNoCache should pass all tests - spinal.core.SpinalExit: 


 Error detected in phase PhaseCheck_noLatchNoOverride
********************************************************************************
********************************************************************************

ASSIGNMENT OVERLAP completely the previous one of (toplevel/io_output : out Bits[8 bits])
    spinal.sim.JvmThread.run(SimManager.scala:51)


********************************************************************************
********************************************************************************

Design's errors are listed above.
SpinalHDL compiler exit stack :
===============================================
```

</details>